velox/exec/HashProbe.cpp

```diff
@@ -220,49 +225,59 @@ void HashProbe::initializeFilter(
-  // the corresponding spilled probe partition on disk.
-  if (restoredPartitionId.has_value()) {
-    auto iter = spillPartitionSet_.find(restoredPartitionId.value());
-    VELOX_CHECK(iter != spillPartitionSet_.end());
-    auto partition = std::move(iter->second);
-    VELOX_CHECK_EQ(partition->id(), restoredPartitionId.value());
-    spillInputReader_ = partition->createUnorderedReader(pool());
-    spillPartitionSet_.erase(iter);
-  }
-
-  VELOX_CHECK_NULL(spiller_);
   spillInputPartitionIds_ = spillPartitionIds;
   if (spillInputPartitionIds_.empty()) {
     return;
   }
 
   // If 'spillInputPartitionIds_' is not empty, then we set up a spiller to
   // spill the incoming probe inputs.
-  const auto& spillConfig = spillConfig_.value();
-  spiller_ = std::make_unique<Spiller>(
+  inputSpiller_ = std::make_unique<Spiller>(
       Spiller::Type::kHashJoinProbe,
       probeType_,
       HashBitRange(
           spillInputPartitionIds_.begin()->partitionBitOffset(),
           spillInputPartitionIds_.begin()->partitionBitOffset() +
-              spillConfig.numPartitionBits),
-      &spillConfig);
+              spillConfig()->numPartitionBits),
+      spillConfig());
   // Set the spill partitions to the corresponding ones at the build side. The
   // hash probe operator itself won't trigger any spilling.
-  spiller_->setPartitionsSpilled(toPartitionNumSet(spillInputPartitionIds_));
+  inputSpiller_->setPartitionsSpilled(
+      toPartitionNumSet(spillInputPartitionIds_));
 
   spillHashFunction_ = std::make_unique<HashPartitionFunction>(
-      spiller_->hashBits(), probeType_, keyChannels_);
+      inputSpiller_->hashBits(), probeType_, keyChannels_);
   spillInputIndicesBuffers_.resize(spillHashFunction_->numPartitions());
   rawSpillInputIndicesBuffers_.resize(spillHashFunction_->numPartitions());
   numSpillInputs_.resize(spillHashFunction_->numPartitions(), 0);
+  // If we have received no more input signal from either source or restored
+  // spill input, then we shall just finish the spiller and records the spilled
+  // partition set accordingly.
+  if (noMoreSpillInput_) {
+    inputSpiller_->finishSpill(spillPartitionSet_);
+  }
+}
+
+void HashProbe::maybeSetupSpillInputReader(
+    const std::optional<SpillPartitionId>& restoredPartitionId) {
+  VELOX_CHECK_NULL(spillInputReader_);
+  if (!restoredPartitionId.has_value()) {
+    return;
+  }
+  // If 'restoredPartitionId' is not null, then 'table_' is built from the
+  // spilled build data. Create an unsorted reader to read the probe inputs from
+  // the corresponding spilled probe partition on disk.
+  auto iter = spillPartitionSet_.find(restoredPartitionId.value());
+  VELOX_CHECK(iter != spillPartitionSet_.end());
+  auto partition = std::move(iter->second);
+  VELOX_CHECK_EQ(partition->id(), restoredPartitionId.value());
+  spillInputReader_ = partition->createUnorderedReader(pool());
+  spillPartitionSet_.erase(iter);
 }
 
 void HashProbe::asyncWaitForHashTable() {
```

Comment:
nit: would be great to update the method comment to mention that it also sets up internal state for potentially spilling input or reading spilled input or recursively spill the hash table.